artiq/devices/novatech409B/novatech409B.py

```diff
@@ -0,0 +1,362 @@
+            self.set_phase(ch_no, phase[ch_no])
+        # send command necessary to update all channels at the same time
+        self.ser_send("I p")
+
+    def freq_sweep_all_phase_continuous(self, f0, f1, t):
+        """ sweep phase of all DDSs, phase continuous
+
+        Sweep frequency in a phase continuous fashion.
+
+        :param float f0: starting frequency (MHz)
+        :param float f1: ending frequency (MHz)
+        :param float t: sweep duration (seconds)
+        :returns: None
+        """
+        s = str(f0) + "," + str(f1) + "," + str(t)
+        self.debug_message("freq_sweep_all_phase_continuous", s, level=2)
+        if f0 == f1:
+            return
+        # get sign of sweep
+        if f1 > f0:
+            df_sign = 1
+        else:
+            df_sign = -1
+
+        self.set_phase_continuous(True)
+        self.set_simultaneous_update(True)
+        # calculate delay
+        # note that a single call to self.set_freq_all_phase_continuous()
+        # takes time t_for_one_freq_set; fix duration empirically
+        t_for_one_freq_set = 0.264
+        dt = t_for_one_freq_set
+        n_steps = int(math.ceil(t/dt))
+        df = abs(f0-f1)/n_steps
+        for n in range(n_steps):
+            fnow = f0+n*df_sign*df
+            self.set_freq_all_phase_continuous(fnow)
+            if self.__debug > 0:
+                print(".", end=" ")
+        self.set_freq_all_phase_continuous(f1)
+
+    def output_scale(self, ch_no, frac):
+        """changes amplitude of a DDS
+
+        :param int ch_no: DDS channel 0, 1, 2 or 3
+        :param float frac: 0 to 1 (full attenuation to no attenuation)
+        :returns: None
+        """
+        self.set_simultaneous_update(False)
+        dac_ch_no = int(math.floor(frac*1024))
+        s = "V{:d} {:d}".format(ch_no, dac_ch_no)
+        self.debug_message("scaleOutput", s, level=3)
+        self.ser_send(s)
+
+    def output_scale_all(self, frac):
+        """changes amplitude of all DDSs
+
+        :param float frac: 0 to 1 (full attenuation to no attenuation)
+        """
+        self.debug_message("scaleOutput", str(frac), level=2)
+        for ch_no in range(4):
+            self.output_scale(ch_no, frac)
+        # send command necessary to update all channels at the same time
+        self.ser_send("I p")
+
+    def output_on_off(self, ch_no, my_bool):
+        """turns on or off the DDS
+
+        :param int ch_no:DDS channel 0, 1, 2 or 3
+        :param bool my_bool: True (if on) or False (if off)
+        """
+        self.debug_message("output_on_off", "", level=3)
+        if my_bool:
+            #turn on output
+            self.output_scale(ch_no, 1.0)
+        else:
+            #turn off output
+            self.output_scale(ch_no, 0.0)
+
+    def output_on_off_all(self, my_bool):
+        """:param bool my_bool: is True (on) or False (off)
+
+        turns on or off the all the DDSs
+        """
+        self.debug_message("output_on_off_all", "", level=2)
```

Comment:
The problem is that several methods do call `set_simultaneous_update()` as a hidden side effect. It is much better to just expose the functionality of the device as faithfully as possible without adding side effects. Then, at a higher level -- e.g. in the experiment code -- you can abstract to compound methods.

More specifically, if I understand this correctly, the correct terminology is transactions. You are in fact opening a transaction with `I m` and committing that transaction with `I p`. And `I a` is implicitly opening a transaction and committing it on every command. You can expose that functionality much more idiomatically.